XInput Tatacon support

endiannessta opened the pull request:
Initial support for XInput Tatacon. Controller mappings intended for PS2 defaults. Has only been tested with the HORI Taiko Drum for PS5, but should be compatible with other XInput controllers. Am just lacking VID & PID's for them.

endiannessta committed:
xinput tatacon support

mappings correspond to defaults in "tatacon de dodon ga don" but should be compatible with any ps2 taiko game. extra vid/pid's would be appreciated.

dinsfire64 commented:
Thank you so much for this PR!

This is looking good, but since I'm not familiar with the Taiko games is there any special pad recognition the games desire to be known as a "special controller"?

Reason I bring it up is Konami games and even the Guitar Hero franchise like to hold Dpad directions to let the game know the "special" controller has connected vs a regular dualshock. You can check the code to see how I keep track of a `special_controller_t` for the current input report.

Does the Taiko franchise do any of this? I presume you are using this on the PS2 games in question as a target.

endiannessta commented:
The original PS2 drums seem to report as a PSX digital pad (confirmed by plugging in a psx controller), and do not have any buttons aside from select and start. Will look to implement this.

dinsfire64 commented:
Yeah it might not identify with anything special at all, but since I'm unfamiliar I thought I would ask.

 If you have a PS2 capable of homebrew, this application is great for testing.
 
 https://www.psx-place.com/resources/ps2-controller-tester-by-jbit.670/

endiannessta committed:
allow tatacon detection in game

very silly controller that acts as a psx digital pad and does not respond to config commands.

endiannessta commented:
game now detects a tatacon successfully. only caveat is that the special device type does not update until you restart the snek.

dinsfire64 commented:
Great! 

I see you are stubbing out some opcodes, can you provide the taiko version you are testing against in game? And how does the game confirm it is a drum or not?

I want to test on my end an idea.

I have a suspicion that it is looking for the digital value of a controller rather than a full dual shock, and I'd rather use that detection method than to stop responding to dualshock opcodes. 

You can see that I did that for the guitar hero games past two. 

https://github.com/dinsfire64/snekbox_firmware/blob/e60dff4600ebde85b96241aed932485ce3a27490/targets/ps2_controller.c#L476

You can also force the isGuitar to true and see if that works as well. We could redefine that variable to be more clear.

endiannessta commented:
Have been testing against Taiko Drum Master, but each PS2 game should detect the same. As far as I'm aware, the only official drums for the system were manufactured by Namco and RedOctane, and they are compatible with every PS2 game in the franchise. There's actually a third party game "Kamen Rider Hibiki" that supports the drum.

According to [https://github.com/PCSX2/pcsx2/issues/4763#issuecomment-2269081492](https://github.com/PCSX2/pcsx2/issues/4763#issuecomment-2269081492), the drum operates as such:
> It's just a ps1 digital pad. Reports as `0x41`, does not support escape/config commands.
> 
> The game even detects a ps1 digital pad as a Taiko controller.
> 
> Buttons: Select, Start.
> 
> And sensors from left to right: L1, Left, Circle, R1

I do agree with you that my methodology was a little uncouth.

dinsfire64 commented:
okay I was able to play around with that game real quick and tried changing a bunch of the constants to attempt to handshake differently with the backend code, but didn't have much success being declared a drum until I took your method of NOPing opcodes.

So the bare minimum this game seems to care about is not responding to the config command. So for example if I do this:
```
    case CMD_CONFIG:
        //processConfig();
        break;
```

The game shows a drum controller in the menu.

If you want to revise your PR to just stub that one command during the taiko drum mode, I think that is a valid compromise.

endiannessta committed:
ignore only the config command when we are a tatacon

endiannessta commented:
I don't believe real digital pads respond to the config command. Playing around with this too. When trying to get the handshake to report a digital pad, the helpful controller tester you recommended will work, but in-game it won't detect a device. For the sake of correctness, I'd like to rework the PS2 target to allow outputting a digital pad, but for now this will do.

dinsfire64 commented:
Yeah if you are happy with the current commit, I can merge and squash. 

And on a separate PR having a "force psx/digital pad" trip would also be useful.

endiannessta commented:
cool to squash. it would be nice in the future to gather the vid/pid's of community made drums as most of them have an xinput mode. the only device this couldn't support is the switch drum, but a handler for it can be worked out

dinsfire64 commented:
Sure thing, I'll do squash and merge this afternoon. 

And yeah having a list of those would be useful, if you drum up interest in that community let me know. 

Apologies for the terrible pun. Haha

endiannessta commented:
rly upsetting that github won't allow me to react with a ba dum tss emoji
